addons/point_of_sale/tests/test_pos_products_with_tax.py

```diff
@@ -541,3 +541,128 @@ def test_fixed_tax_positive_qty(self):
+        tax_xx = self.env['account.tax'].create({
+            'name': 'Tax XX',
+            'type_tax_use': 'sale',
+            'amount_type': 'percent',
+            'amount': 25,
+            'tax_group_id': tax_groups[2].id,
+            'company_id': branch_xx.id,
+        })
+        # create several products with different taxes combination
+        product_all_taxes = self.env['product.product'].create({
+            'name': 'Product all taxes',
+            'available_in_pos': True,
+            'taxes_id': [odoo.Command.set((tax_a + tax_b + tax_x + tax_xx).ids)],
+        })
+        product_no_xx_tax = self.env['product.product'].create({
+            'name': 'Product no tax from XX',
+            'available_in_pos': True,
+            'taxes_id': [odoo.Command.set((tax_a + tax_b + tax_x).ids)],
+        })
+        product_no_branch_tax = self.env['product.product'].create({
+            'name': 'Product no tax from branch',
+            'available_in_pos': True,
+            'taxes_id': [odoo.Command.set((tax_a + tax_b).ids)],
+        })
+        product_no_tax = self.env['product.product'].create({
+            'name': 'Product no tax',
+            'available_in_pos': True,
+            'taxes_id': [],
+        })
+        # configure a session on Branch XX
+        xx_config = self.env['pos.config'].with_company(branch_xx).create({
+            'name': 'Branch XX config',
+            'company_id': branch_xx.id,
+        })
+        xx_account_receivable = self.company_data['default_account_receivable'].copy()
+        xx_account_receivable.company_id = branch_xx
+        xx_cash_journal = self.company_data['default_journal_cash'].copy()
+        xx_cash_journal.company_id = branch_xx
+        xx_cash_payment_method = self.env['pos.payment.method'].create({
+            'name': 'XX Cash Payment',
+            'receivable_account_id': xx_account_receivable.id,
+            'journal_id': xx_cash_journal.id,
+            'company_id': branch_xx.id,
+        })
+        xx_config.write({'payment_method_ids': [
+            odoo.Command.set(xx_cash_payment_method.ids),
+        ]})
+        self.config = xx_config
+        pos_session = self.open_new_session()
+        # load the session data from Branch XX:
+        # - Product all taxes           => tax from Branch XX should be set
+        # - Product no tax from XX      => tax from Branch X should be set
+        # - Product no tax from branch  => 2 taxes from parent company should be set
+        # - Product no tax              => no tax should be set
+        pos_data = pos_session.load_pos_data()
+        self.assertEqual(
+            next(iter(filter(lambda p: p['id'] == product_all_taxes.id, pos_data['product.product'])))['taxes_id'],
+            tax_xx.ids
+        )
+        self.assertEqual(
+            next(iter(filter(lambda p: p['id'] == product_no_xx_tax.id, pos_data['product.product'])))['taxes_id'],
+            tax_x.ids
+        )
+        self.assertEqual(
```

Comment:
I do not know if there is a tiny probability that `next(iter(filter(lambda p: p['id'] == product_all_taxes.id, pos_data['product.product'])))['taxes_id']` returns an unsorted array but if it does, the assertEqual will throw an AssertionError so maybe a quick sorting here might be good to assure that the test pass could be good. I'll let you decide whether or not it is necessary.

Reply:
Indeed. It could happen.
I've added a `sort()` on the assert where there are 2 taxes.